### PR TITLE
Avoid deep copy in ``DataFrame.__setitem__`` for pandas 1.4 and up

### PR DESCRIPTION
**Pull request by phofl**
- [ ] Closes #xxxx
- [ ] Tests added / passed
- [ ] Passes `pre-commit run --all-files`

cc @fjetter 

This should be an easy win

### CODE REVIEW

File: dask/dataframe/methods.py [hunk 1]
```diff
@@ -340,8 +341,9 @@ def cummax_aggregate(x, y):
 def assign(df, *pairs):
     # Only deep copy when updating an element
     # (to avoid modifying the original)
+    # Setitem never modifies an array inplace with pandas 1.4 and up
     pairs = dict(partition(2, pairs))
-    deep = bool(set(pairs) & set(df.columns))
+    deep = bool(set(pairs) & set(df.columns)) and not PANDAS_GT_140
     df = df.copy(deep=bool(deep))
     for name, val in pairs.items():
         df[name] = val
```
Comment (fjetter) [inline dask/dataframe/methods.py:349]: Why can't we just do `df = df.assign(pairs)`? I would assume this to be handled better by pandas.

File: dask/dataframe/methods.py [hunk 2]
```diff
@@ -340,8 +341,9 @@ def cummax_aggregate(x, y):
 def assign(df, *pairs):
     # Only deep copy when updating an element
     # (to avoid modifying the original)
+    # Setitem never modifies an array inplace with pandas 1.4 and up
```
Comment (fjetter) [inline dask/dataframe/methods.py:344]: instead every assign will create a copy?

File: dask/dataframe/methods.py (hunk 1 above)
Comment (phofl) [reply dask/dataframe/methods.py:349]: assign copies the data while ``df[x] = something`` doesn't

**Review by fjetter — APPROVED** [review]

**Review by jrbourbeau — COMMENTED** [review]
Thanks @phofl @fjetter. Is there some sort of test we could reasonably add for this? I could imagine someone through in the future rewriting things and, unintentionally, changing this subtle behavior. (I'm thinking less about this specific change, and more about if we end up making several similar updates)

**Comment by phofl** [comment]
I'll add a test

**Pull request closed without merging**